Research/lva-node-app/public/js/mediaGraphBuilder.js

```diff
@@ -0,0 +1,1114 @@
+//media graph building related functions
+/**
+ * @const GITHUB_TOPOLOGY_SAMPLES - link to raw github where media graph samples exist
+ */
+const GITHUB_TOPOLOGY_SAMPLES = 'https://raw.githubusercontent.com/Azure/live-video-analytics/master/MediaGraph/topologies/';
+var sourceNames = [];
+var processorNames = [];
+var sinkNames = [];
+
+/**
+* class for a graph node. Dynamically built upon media graph creation, based off the node type and the 
+* inputs. 
+@class
+*/
+class GraphNode {
+
+    /**
+     * @constructor
+     * @param {string} type - The type of node to create
+     * @param {list of strings} userSelectedInputs - The list of inputs the user selected for this graph node
+     */
+    constructor(type, userSelectedInputs)
+    {
+        //ensure node type exists
+        this.nodeSchema=getNodeSchema(type);
+        if (this.nodeSchema==undefined)
+        {
+            throw ("Node type: "+type+" not found");
+        } 
+
+        //set inputs
+        this.inputs=[];
+        for(let i=0; i<userSelectedInputs.length; i++)
+        {
+            if (findObjectByKey(this.nodeSchema.inputs, "name", userSelectedInputs[i]))
+            {
+                this.inputs.push({"nodeName": userSelectedInputs[i]});
+            }
+            else
+            {
+                throw ("Input "+userSelectedInputs[i]+" is not accepted by node type "+type);
+            } 
+        }
+
+        this.jsonObj = 
+        {
+            "name": type,
+            "@type": this.nodeSchema.type,
+        }
+
+        // must check that not a source node. Otherwise, will break media graph
+        if (!type.toLowerCase().includes("source"))
+        {
+            this.jsonObj["inputs"]=this.inputs;
+        }
+
+        for(let i=0; i<this.nodeSchema.extraValues.length; i++)
+        {
+            const item=this.nodeSchema.extraValues[i];
+            this.jsonObj[item.name]=item.value;
+        }
+        //we have now constructed the full json object!
+
+        this.nodeParameters=this.nodeSchema.extraParameters;
+    }
+}
+
+/**
+ * MediaGraph class, contains name, desciption, sources, processors, sinks, and parameters. Creates a nice jsonObject to be used as a payload later. 
+ * @class
+ */
+class MediaGraph
+{
+    /**
+     * @constructor
+     * @param {string} graphname 
+     * @param {string} description 
+     * @param {list of GraphNodes} sources 
+     * @param {list of GraphNodes} processors 
+     * @param {list of GraphNodes} sinks 
+     * @param {list of Objects containing parameter info} parameters 
+     */
+    constructor(graphname, description="no description set", sources, processors, sinks, parameters)
+    {
+        this.graphname=graphname;
+        this.sources=sources;
+        this.processors=processors;
+        this.sinks=sinks;
+        this.parameters=parameters;
+        this.jsonObject = {
```

Comment:
this seems like it should be created dynamically when needed. otherwise if at some point the code reasonably changes one of the class vars (ex sinks), the jsonObject will still be on the old setting.

Reply:
you would never modify an existing instance of type MediaGraph. For this reason, I think you are right to suggest that we get rid of the MediaGraph class and just have a function that creates the jsonObj since that is all that we ever use! I've implemented these changes and will push shortly